src/main/clojure/clara/rules/engine.cljc

```diff
@@ -1784,52 +1778,100 @@
+        ;;
+        ;; We perform the insertions and retractions in the same order as they were applied to the session since
+        ;; if a fact is not in the session, retracted, and then subsequently inserted it should be in the session at
+        ;; the end.
+        (do
+          (doseq [{op-type :type facts :facts} pending-operations]
 
-            (binding [*pending-external-retractions* (atom facts)]
-              (external-retract-loop get-alphas-fn transient-memory transport transient-listener)))))
+            (case op-type
 
-      (fire-rules* rulebase
-                   (:production-nodes rulebase)
-                   transient-memory
-                   transport
-                   transient-listener
-                   get-alphas-fn)
+              :insertion
+              (do
+                (l/insert-facts! transient-listener facts)
+
+                (binding [*pending-external-retractions* (atom [])]
+                  ;; Bind the external retractions cache so that any logical retractions as a result
+                  ;; of these insertions can be cached and executed as a batch instead of eagerly realizing
+                  ;; them.  An external insertion of a fact that matches
+                  ;; a negation or accumulator condition can cause logical retractions.
+                  (doseq [[alpha-roots fact-group] (get-alphas-fn facts)
+                          root alpha-roots]
+                    (alpha-activate root fact-group transient-memory transport transient-listener))
+                  (external-retract-loop get-alphas-fn transient-memory transport transient-listener)))
+
+              :retraction
+              (do
+                (l/retract-facts! transient-listener facts)
+
+                (binding [*pending-external-retractions* (atom facts)]
+                  (external-retract-loop get-alphas-fn transient-memory transport transient-listener)))))
+
+          (fire-rules* rulebase
+                       (:production-nodes rulebase)
+                       transient-memory
+                       transport
+                       transient-listener
+                       get-alphas-fn
+                       (uc/get-ordered-update-cache)))
+
+        #?(:cljs (throw (ex-info "Using the option for insertions and retractions to cancel each other out is not supported in ClojureScript"
+                                 {:session session :opts opts}))
+
+           :clj (do (let [insertions (into []
+                                           (comp (filter (fn [pending-op]
+                                                           (= (:type pending-op)
+                                                              :insertion)))
+                                                 (map :facts)
+                                                 cat)
+                                           pending-operations)
+                          retractions (into []
```

Comment:
Minor: when building "throw away" collections that need a transformation like this, building a vector is heavier than necessary .  `sequence` is a good alternative here.